badge-maker/lib/xml.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * @module
+ */
+
+'use strict'
+
+function stripXmlWhitespace(xml) {
+  return xml.replace(/>\s+/g, '>').replace(/<\s+/g, '<').trim()
+}
+
+function escapeXml(s) {
+  if (typeof s === 'number') {
+    return s
+  } else if (s === undefined || typeof s !== 'string') {
+    return undefined
+  } else {
+    return s
+      .replace(/&/g, '&amp;')
+      .replace(/</g, '&lt;')
+      .replace(/>/g, '&gt;')
+      .replace(/"/g, '&quot;')
+      .replace(/'/g, '&apos;')
+  }
+}
+
+/**
+ * Representation of an XML element
+ */
+class XmlElement {
+  /**
+   * Xml Element Constructor
+   *
+   * @param {object} attrs Refer to individual attrs
+   * @param {string} attrs.tag
+   *    Name of the XML tag
+   * @param {Array.<string|module:badge-maker/lib/xml-element~XmlElement>} [attrs.content=[]]
+   *    Array of objects to render inside the tag. content may contain a mix of
+   *    string and XmlElement objects. If content is `[]` or ommitted the
+   *    element will be rendered as a self-closing element.
+   * @param {object} [attrs.attrs={}]
+   *    Object representing the tag's attributes as name/value pairs
+   */
+  constructor({ tag, content = [], attrs = {} }) {
+    this.tag = tag
+    this.content = content
+    this.attrs = attrs
+  }
+
+  /**
+   * Render the XML element to a string, applying appropriate escaping
+   *
+   * @returns {string} String representation of the XML element
+   */
+  render() {
+    let attrsStr = ''
+    attrsStr = Object.entries(this.attrs)
+      .map(([k, v]) => `${k}="${escapeXml(v)}"`)
+      .join(' ')
+    if (this.content.length > 0) {
+      let xmlStr = `<${this.tag}`
+      if (attrsStr === '') {
+        xmlStr += '>'
+      } else {
+        xmlStr += ` ${attrsStr}>`
```

Comment:
admittedly this may be another comment in the arena of micro-optimization, but would it be easier to just always insert the extra space (the else path), or if we want to avoid having a ` >`, add an inline ternary on the end of the initial `xmlStr` assignment? I got a bit lost in the preceding conversation so while I understand and agree the useless space is indeed undesired, I wasn't clear whether we've already got cases of that today.

Just wondering if it'd be beneficial to be able to avoid the string concatenation since this will be on a hot path with recursion

Reply:
I suspect [it doesn't matter](https://blog.codinghorror.com/the-sad-tragedy-of-micro-optimization-theater/)

While conceptually this is recursive, for a badge the deepest trees we create are like 5 nodes deep so.. meh

I've pushed a version with fewer template statements, fewer concatenations and zero mutable variables. I won't make any claims about performance, but having revisited it, I reckon this is at least easier to read, so that's something.